sysl2/sysl/deps.go

```diff
@@ -0,0 +1,264 @@
+	s := StrSet{}
+	s.Insert(dep.Self.Endpoint)
+	s.Insert(dep.Target.Endpoint)
+	return s
+}
+
+func NewDependencySet() *DependencySet {
+	return &DependencySet{map[string]*AppDependency{}}
+}
+
+func MakeAppDependency(self, target *AppElement, stmt *sysl.Statement) *AppDependency {
+	return &AppDependency{self, target, stmt}
+}
+
+func MakeAppElement(name, endpoint string) *AppElement {
+	return &AppElement{name, endpoint}
+}
+
+type CallStatementSlice struct {
+	slice []*sysl.Statement
+}
+
+func NewCallStatementSlice() *CallStatementSlice {
+	return &CallStatementSlice{}
+}
+
+func (cs *CallStatementSlice) Add(call *sysl.Statement) {
+	cs.slice = append(cs.slice, call)
+}
+
+func (cs *CallStatementSlice) GetSlice() []*sysl.Statement {
+	return cs.slice
+}
+
+func (cs *CallStatementSlice) CollectCallStatements(stmts []*sysl.Statement) {
+	for _, stat := range stmts {
+		switch c := stat.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			cs.Add(stat)
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			cs.CollectCallStatements(c.Cond.GetStmt())
+		case *sysl.Statement_Loop:
+			cs.CollectCallStatements(c.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			cs.CollectCallStatements(c.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			cs.CollectCallStatements(c.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			cs.CollectCallStatements(c.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range c.Alt.GetChoice() {
+				cs.CollectCallStatements(choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func FindApps(module *sysl.Module, excludes, integrations StrSet, ds []*AppDependency, highlight bool) StrSet {
+	r := []string{}
+	appReStr := toPattern(integrations.ToSlice())
+	re := regexp.MustCompile(appReStr)
+	st, _ := stream.Of(ds)
+	st.ForEach(func(v stream.T) bool {
+		if dep, ok := v.(*AppDependency); ok {
+			appNames := dep.extractAppNames()
+			excludeApps := appNames.Intersection(excludes)
+			if len(excludeApps) > 0 {
+				return true
+			}
+			highlightApps := appNames.Intersection(integrations)
+			if !highlight && len(highlightApps) == 0 {
+				return true
+			}
+			appStream, _ := stream.Of(appNames.ToSlice())
+			appStream.ForEach(func(v stream.T) bool {
+				item := v.(string)
+				app := module.GetApps()[item]
+				if highlight {
+					if re.MatchString(item) &&
+						!hasPattern("human", app.GetAttrs()) {
+						r = append(r, item)
+					}
+					return true
+				}
+				if !hasPattern("human", app.GetAttrs()) {
+					r = append(r, item)
+				}
+				return true
+			})
+		}
+		return true
+	})
+
+	return MakeStrSet(r...)
+}
+
+func walkPassthrough(excludes, passthroughs StrSet, dep *AppDependency, integrations *DependencySet, module *sysl.Module) {
+	target := dep.Target
+	targetName := target.Name
+	targetEndpoint := target.Endpoint
+	excludedApps := MakeStrSet(targetName).Intersection(excludes)
```

Comment:
@JackLeeHal remove target, targetName and targetEndpoint. use the actual variables directly.

Reply:
removed them.